syntax = "proto3";

import "google/protobuf/timestamp.proto";


message ProjectCreated{
    string project_id = 1;
    string name = 2;
}

message DatasetCreated{
    string dataset_id = 1;
    string project_id = 2;
    string name = 3;
    DatasetMetadata meta = 4;
}

message DatasetUpdated {
    string dataset_id = 1;
    string project_id = 2;
    DatasetMetadata meta = 3;
}



message JobAdded {
    string job_id = 1;
    string job_name = 2;

    repeated string inputs = 3;
    repeated string outputs = 4;

    string project_id = 5;


}


message ExpertAdded {
    string expert_id = 1;
    string expert_name = 2;
}


enum STATE {
    NONE = 0;
    VALUE = 1;
    DELETE = 2;
}





message DatasetSample {
    // Sample storage format. Doesn't necessarily match
    // actual data format which could be very cryptic
    enum FORMAT {
        TEXT = 0;
        TSV = 1;
        JSON = 2;
    }

    FORMAT format = 1;
    bytes body = 2;
}


// if a field kind is set, then the corresponding property carries
// a new value (which could be null)
message DatasetMetadata {

    int64 record_count = 1;
    STATE record_count_state = 2;
    int64 file_count = 3;
    STATE file_count_state = 4;
    int64 storage_bytes = 5;
    STATE storage_bytes_state = 6;
    DatasetSample sample = 7;
    STATE sample_state = 8;

    int64 update_timestamp = 9;
    STATE update_timestamp_state = 10;

    string data_format = 11;
    STATE data_format_state = 12;
    string description = 13;
    STATE description_state = 14;

    string location_id = 15;
    STATE location_id_state = 16;
    string location_uri = 17;
    STATE location_uri_state=18;


    repeated string experts = 19;
    STATE experts_state = 20;
}

enum Type {
    None = 0;
    Event_ProjectCreated = 1;
    Event_DatasetCreated = 2;
    Event_DatasetUpdated = 3;
    Event_ExpertAdded = 4;
    Event_JobAdded = 5;
}


message Event {
    bytes Body = 1;
    Type Type = 2;
}


